File: ga/4omini_result_2024_7/result/1968F_cpp_mini/7_1968F.cpp
use std::io::{self, BufRead};
use std::collections::BinaryHeap;

fn main() {
    let stdin = io::stdin();
    let reader = stdin.lock();
    let mut lines = reader.lines();
    
    let t: usize = lines.next().unwrap().unwrap().trim().parse().unwrap();
    for _ in 0..t {
        solve(&mut lines);
    }
}

fn solve<R: BufRead>(lines: &mut R) {
    let first_line = lines.next().unwrap().unwrap();
    let mut nums = first_line.trim().split_whitespace();
    
    let n: usize = nums.next().unwrap().parse().unwrap();
    let q: usize = nums.next().unwrap().parse().unwrap();
    
    let mut arr = Vec::with_capacity(n);
    let arr_line = lines.next().unwrap().unwrap();
    
    arr.extend(arr_line.trim().split_whitespace().map(|s| s.parse::<i64>().unwrap()));

    // Create prefix XOR and index pairs
    let mut x = vec![(0, 0); n + 1];
    let mut x2 = vec![0; n + 1];
    
    for i in 1..=n {
        x[i].0 = x[i - 1].0 ^ arr[i - 1];
        x[i].1 = i as i64;
    }
    
    for i in 0..=n {
        x2[i] = x[i].0;
    }
    
    // Sort the x array based on the first element of the pairs
    x.sort();

    for _ in 0..q {
        let query_line = lines.next().unwrap().unwrap();
        let mut query_nums = query_line.trim().split_whitespace();
        
        let l: usize = query_nums.next().unwrap().parse().unwrap();
        let r: usize = query_nums.next().unwrap().parse().unwrap();

        let c = x2[r] ^ x2[l - 1];
        if c == 0 {
            println!("YES");
            continue;
        }
        
        let target = c ^ x2[l - 1];

        // Binary search for i1
        let mut i1 = x.binary_search(&(target, 0)).unwrap_or_else(|e| e);
        
        if i1 == n + 1 || x[i1].1 > r as i64 || x[i1].0 != target {
            println!("NO");
            continue;
        }
        
        // Binary search for i2
        let mut i2 = x.binary_search(&(x2[l - 1], 0)).unwrap_or_else(|e| e);
        
        if i2 == n + 1 || x[i2].1 > r as i64 || x[i2].0 != x2[l - 1] {
            println!("NO");
        } else {
            println!("YES");
        }
    }
}

File: ga/glm4_result_2024_6/result/1968F_cpp_glm/7_1968F.rs
use std::io;
use std::io::{BufRead, Write};

fn solve() {
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let mut input = input.trim().split_whitespace();
    let n: usize = input.next().unwrap().parse().unwrap();
    let q: usize = input.next().unwrap().parse().unwrap();

    let mut arr = Vec::with_capacity(n);
    for _ in 0..n {
        let mut input = String::new();
        io::stdin().read_line(&mut input).unwrap();
        let num: u32 = input.trim().parse().unwrap();
        arr.push(num);
    }

    let mut x = Vec::with_capacity(n + 1);
    let mut x2 = Vec::with_capacity(n + 1);
    x2.push(0);
    for i in 1..=n {
        x2.push(x2[i - 1] ^ arr[i - 1]);
        x.push((x2[i], i));
    }
    x.sort();

    for _ in 0..q {
        let mut input = String::new();
        io::stdin().read_line(&mut input).unwrap();
        let mut input = input.trim().split_whitespace();
        let l: usize = input.next().unwrap().parse().unwrap();
        let r: usize = input.next().unwrap().parse().unwrap();

        let c = x2[r] ^ x2[l - 1];
        if c == 0 {
            println!("YES");
            continue;
        }

        let i1 = match x.binary_search_by_key(&c ^ x2[l - 1], |&(k, _)| k) {
            Ok(i) => i,
            Err(i) => i,
        };

        if i1 == n + 1 || x[i1].1 > r || x[i1].0 != (c ^ x2[l - 1]) {
            println!("NO");
            continue;
        }

        let i2 = match x.binary_search_by_key(&x2[l - 1], |&(k, _)| k) {
            Ok(i) => i,
            Err(i) => i,
        };

        if i2 == n + 1 || x[i2].1 > r || x[i2].0 != x2[l - 1] {
            println!("NO");
        } else {
            println!("YES");
        }
    }
}

fn main() {
    let t: usize = read_int();
    for _ in 0..t {
        solve();
    }
}

fn read_int() -> usize {
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    input.trim().parse().unwrap()
}